spyder/plugins/run/api.py

```diff
@@ -0,0 +1,595 @@
+
+
+class RunResultError(TypedDict):
+    # Error code for the current error.
+    code: Optional[int]
+    # Human-readable message that describes the error.
+    message: Optional[str]
+
+
+class RunResult(TypedDict):
+    """Run result entry schema."""
+
+    # The output format for the current run result. It must belong to the
+    # `RunResultDisplay` dict.
+    output_format: str
+
+    # The output of the run result. The viewer is responsible for the current
+    # interpretation of the output type.
+    run_output: Union[Any, RunResultError]
+
+    # File extension or identifier of the input context. It must belong to the
+    # `RunInputExtension` set.
+    input_extension: str
+
+    # Original run input metadata.
+    metadata: RunConfigurationMetadata
+
+    # Execution metadata.
+    execution_metadata: RunExecutionMetadata
+
+    # True if the execution result yielded an error, False otherwise. If the
+    # execution failed, then the `run_output` field will comply with the
+    # `RunResultError` schema.
+    is_error: bool
+
+
+PossibleRunResult = Union[RunResult, RunResultError]
+
+
+class Context(TypedDict):
+    """Run context name schema."""
+    # CamelCase name of the context.
+    name: str
+    # String identifier for the run context. If non-existent or None, then
+    # a snake_case version of the name is used.
+    identifier: NotRequired[Optional[str]]
+
+
+class OutputFormat(TypedDict):
+    """Output format information schema."""
+    # Human-readable name for the output format. It must be CamelCase
+    name: str
+    # String identifier for the output format. If non-existent or None, then
+    # a snake_case version of the name is used.
+    identifier: NotRequired[Optional[str]]
+
+
+class SupportedRunConfiguration(TypedDict):
+    """Run configuration entry schema."""
+
+    # File extension or identifier of the input context. It must belong to the
+    # `RunInputExtension` set.
+    input_extension: str
+
+    # The supported contexts for the given input extension.
+    # e.g., file, selection, cell, others.
+    # The context can be compared against the values of `RunContext`. e.g.,
+    # `info['context'] == RunContext.File`
+    contexts: List[Context]
+
+
+class WorkingDirSource:
+    ConfigurationDirectory = 0
+    CurrentDirectory = 1
+    CustomDirectory = 2
+
+
+class WorkingDirOpts(TypedDict):
+    """Run execution working directory options."""
+
+    # Source used to look for the working directory that will be used in
+    # the run execution.
+    source: WorkingDirSource
+
+    # If not None, then it specifies a path to a custom directory location.
+    path: Optional[str]
+
+
+class RunExecutionParameters(TypedDict):
+    """Run execution parameters."""
+
+    # The working directory options
+    working_dir: WorkingDirOpts
+
+    # The run executor parameters
+    executor_params: dict
+
+
+class ExtendedRunExecutionParameters(TypedDict):
+    """Extended run execution parameters."""
```

Comment:
Seems like this is acting like the `RunExecutionParameters` metadata but having a reference to its parent instead of being referenced by it:

`ExtendedRunExecutionParameters` -> `RunExecutionParameters` 

So seems like this relation has an inverse structure in comparison with other definitions like for example:

`RunConfiguration` -> `RunConfigurationMetadata` 

Is there a reason for doing this instead of defining the relation as `RunExecutionParameters` ->`RunExecutionParametersMetadata` or something equivalent using the concept of `*Metadata`?

Reply:
They are different, the main goal of a `ExtendedRunExecutionParameters` is to define the `RunExecutionParameters` and extra information needed at the time of execution. For the rest, the `*Metadata` is for registering information in the run plugin that will be retrieved when anything is executed (this one persists in the session while the actual thing that gets executed may change). 

The extended parameters contain different information about the executor parameters (like the external console has multiple shells and options different from the ipython console). The idea is that we have a base of what an executor should provide called `RunExecutionParameters`, and if you need extra of that then this class will be used.